### PR TITLE
Add heuristic default block sizes for different cases in ragged attention kernel

### PR DESCRIPTION
**Pull request by yaochengji**

### CODE REVIEW

File: test/test_pallas.py [hunk 1]
```diff
@@ -856,6 +867,22 @@ def test_ragged_paged_attention_wrapper_without_dynamo(
         use_dynamo=False,
     )
```
Comment (bythew3i) [inline test/test_pallas.py:868]: Remove the old _test_ragged_paged_attention?

File: test/test_pallas.py [hunk 2]
```diff
@@ -817,6 +812,22 @@ def test_ragged_paged_attention_wrapper_with_dynamo(
         use_dynamo=True,
```
Comment (bythew3i) [inline test/test_pallas.py:812]: Remove the old _test_ragged_paged_attention?

File: torch_xla/experimental/custom_kernel.py [hunk 3]
```diff
@@ -955,6 +955,28 @@ def ragged_paged_attention(
...
+      num_kv_pages_per_block = 128
+      num_queries_per_block = 96
+
+  if vmem_limit_bytes is None:
+    tpu_version = torch_xla.tpu.version()
+    if tpu_version < 4:
+        raise NotImplementedError("TPU version must be 4 or higher.")
+    # NOTE: the TPU v4's vmem capacity is 16MB
+    if tpu_version == 4:
+        vmem_limit_bytes = 16 * 1024 * 1024
```
Comment (bythew3i) [inline torch_xla/experimental/custom_kernel.py:976]: It is fine, even we set more than 16MB, it will still use 16 MB

File: torch_xla/experimental/custom_kernel.py [hunk 4]
```diff
@@ -955,6 +955,28 @@ def ragged_paged_attention(
   # in the global scope which could cause problems for xmp.spawn.
   from torch_xla.experimental.pallas_kernels.ragged_paged_attention_v2 import ragged_paged_attention as ragged_attention
 
+  if num_kv_pages_per_block is None and num_queries_per_block is None:
+    token_num = q.shape[0]
+    # This heristic is based on the initial kernel micro benchmarking:
+    # When the token_num is small, there's no long request of prefill.
+    # While when it's larger, the block size is adjusted for it.
+    if token_num <= 128:
```
Comment (vanbasten23) [inline torch_xla/experimental/custom_kernel.py:963]: I wonder if we should choose the block sizes in vLLM. If we choose to do in torch_xla, then we need to change it in torch_xla and wait for the wheel tmr. If we do in vLLM, it'd be more convenient. wdyt?

File: test/test_pallas.py (hunk 1 above)
Comment (yaochengji) [reply test/test_pallas.py]: We'd better also test the non-None block size parameter, that's why it's still there.

File: test/test_pallas.py (hunk 2 above)
Comment (yaochengji) [reply test/test_pallas.py:820]: We'd better also test the non-None block size parameter, that's why it's still there.

File: torch_xla/experimental/custom_kernel.py (hunk 3 above)
Comment (yaochengji) [reply torch_xla/experimental/custom_kernel.py]: Thanks, done.

File: test/test_pallas.py [hunk 5]
```diff
@@ -817,6 +812,22 @@ def test_ragged_paged_attention_wrapper_with_dynamo(
...
+        head_dim,
+        page_size,
+        num_pages,
+        dtype,
+        sm_scale=sm_scale,
+        sliding_window=sliding_window,
+        soft_cap=soft_cap,
+        pad_tokens_and_seqs=pad_tokens_and_seqs,
+        use_dynamo=True,
+        num_kv_pages_per_block=None,
```
[the remark this answers is not in the document]
Comment (yaochengji) [reply test/test_pallas.py]: Thanks for the suggestion, done.

File: torch_xla/experimental/custom_kernel.py (hunk 4 above)
Comment (yaochengji) [reply torch_xla/experimental/custom_kernel.py]: If vLLM pass a non-None block size, the default value will not be used.

Comment (vanbasten23) [reply torch_xla/experimental/custom_kernel.py]: I wonder why we couldn't do `_get_default_ragged_paged_attention_block_size` in vLLM..

Comment (yaochengji) [reply torch_xla/experimental/custom_kernel.py]: Usually it's a good idea to put the tuned-parameter table in the kernel lib, not in the app lib.

**Review by vanbasten23 — APPROVED** [review]
Thanks Chengji!
